File: Python/Hard/SlidingWindowMaximum.py
from typing import List
from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        result = []
        dq = deque()

        for i in range(k):

            while dq and nums[dq[-1]] < nums[i]:
                dq.pop()
            dq.append(i)

        result.append(nums[dq[0]])

        for i in range(k, len(nums)):

            while dq and dq[0] <= i - k:
                dq.popleft()

            while dq and nums[dq[-1]] < nums[i]:
                dq.pop()

            dq.append(i)
            result.append(nums[dq[0]])

        return result

File: JavaScript/Hard/SlidingWindowMaximum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function (nums, k) {
  const result = [];
  const deque = [];

  for (let i = 0; i < k; i++) {
    while (deque.length && nums[deque[deque.length - 1]] < nums[i]) {
      deque.pop();
    }
    deque.push(i);
  }

  result.push(nums[deque[0]]);

  for (let i = k; i < nums.length; i++) {
    while (deque.length && deque[0] <= i - k) {
      deque.shift();
    }

    while (deque.length && nums[deque[deque.length - 1]] < nums[i]) {
      deque.pop();
    }

    deque.push(i);
    result.push(nums[deque[0]]);
  }

  return result;
};
